src/main.rs

```diff
@@ -1,47 +1,216 @@
+#![allow(clippy::unusual_byte_groupings)]
+#![allow(clippy::upper_case_acronyms)]
 use std::fs;
 
 const MEMORY_SIZE: usize = 2_usize.pow(16);
 
+#[derive(Debug)]
+enum Argument {
+    Reg(u16),
+    Immediate(i16),
+}
+
 #[derive(Debug)]
 enum Opcode {
-    BR, // Branch
-    ADD, // Add
-    LD, // Load
-    ST, // Store
-    JSR, // Jump Register
-    AND, // Bitwise And
-    LDR, // Load Register
-    STR, // Store Register
-    RTI, // Unused
-    NOT, // Bitwise Not
-    RET, // Return from JSR
-    LDI, // Load Indirect
-    STI, // Store Indirect
-    JMP, // Jump
-    RES, // Unused
-    LEA, // Load Effective Address
-    TRAP // Execute Trap
+    ADD {
+        dr: u16,
+        sr1: u16,
+        sr2: Argument,
+    }, // Add (dr <- sr1 + sr2/imm5)
+    AND {
+        dr: u16,
+        sr1: u16,
+        sr2: Argument,
+    }, // Bitwise And (dr <- sr1 & sr2/imm5)
+    BR {
+        n: bool,
+        z: bool,
+        p: bool,
+        offset: i16,
+    }, // Branch (If n/z/p => pc += offset)
+    JMP {
+        base_r: u16,
+    }, // Jump (pc <- base_r)
+    RET, // Return from subroutine (JMP) (pc <- r7)
+    JSR {
+        offset: i16,
+    }, // Jump To Subroutine (pc <- pc + offset)
+    JSRR, // Jump to Subroutine (pc <- base_r)
+    LD {
+        dr: u16,
+        offset: i16,
+    }, // Load Direct from memory (dr <- mem[pc+offset])
+    LDI {
+        dr: u16,
+        offset: i16,
+    }, // Load Indirect from memory (dr <- mem[mem[pc+offset]])
+    LDR {
+        dr: u16,
+        base_r: u16,
+        offset: i16,
+    }, // Load Base+Offset (dr <- mem[base_r + offset])
+    LEA {
+        dr: u16,
+        offset: i16,
+    }, // Load Effective Address (dr <- pc + offset)
+    NOT {
+        dr: u16,
+        sr: u16,
+    }, // Bitwise Not (dr <- ~sr)
+    RTI, // Return From Interrupt
+    ST {
+        sr: u16,
+        offset: i16,
+    }, // Store (mem[pc+offset] <- sr)
+    STI {
+        sr: u16,
+        offset: i16,
+    }, // Store Indirect (mem[mem[pc+offset]] <- sr)
+    STR {
+        sr: u16,
+        base_r: u16,
+        offset: i16,
+    }, // Store Base+Offset (mem[sr+offset] <- sr)
+    TRAP, // Execute Trap
+    RESERVED, // Unused. Throws an Illegal Opcode Exception.
 }
 
-// 1010110101010111
 fn decode_instruction(instruction: u16) -> Option<Opcode> {
     let op = (instruction >> 12) as u8; // Highest 4 bits
     match op {
-        0b0001 => Some(Opcode::ADD),
-        0b0101 => Some(Opcode::AND),
-        0b0000 => Some(Opcode::BR),
-        0b1100 => Some(Opcode::JMP),
-        0b0100 => Some(Opcode::JSR),
-        0b0010 => Some(Opcode::LD),
-        0b1010 => Some(Opcode::LDI),
-        0b0110 => Some(Opcode::LDR),
-        0b1110 => Some(Opcode::LEA),
-        0b1001 => Some(Opcode::NOT),
+        0b0001 => {
```

Comment:
Could this bytes be put into constants?

Like 

```rust
const ADD: byte =  0b0001;
```

What do you think?

Reply:
Agreed

Reply:
They are only used there and they are pretty clear from the specification that those are the bits that determine the opcode; this would imply adding many constants just for 1 use each.

Besides, if one of the constants is badly written e.g.:
`const JMP: byte = 0b1101`
it's harder to debug if it's somewhere else.

Reply:
> They are only used there and they are pretty clear from the specification that those are the bits that determine the opcode; this would imply adding many constants just for 1 use each.
> 
> Besides, if one of the constants is badly written e.g.: `const JMP: byte = 0b1101` it's harder to debug if it's somewhere else.

More than fair point!